### PR TITLE
feat: implement channels:configure command

### PR DESCRIPTION
**Pull request by oliver-daniel**
~~Still awaiting tests.~~

### CODE REVIEW

File: src/commands/channels/configure.ts [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    const modesSeen = Object.fromEntries(
+      ChannelModalities
+        .slice(1)
+        .map((mode: string) => [mode, false]))
+
+    const modesOptions = options.mode?.map((mode: string) => mode.toUpperCase())
+
+    for (const mode of modesOptions ?? []) {
+      if (!(mode in modesSeen)) {
+        // TODO
```
Comment (oliver-daniel) [inline src/commands/channels/configure.ts:84]: Left as TODO because I want to talk about whether more descriptive messages are needed.

File: src/mix/api/channels-types.ts [hunk 2]
```diff
@@ -21,4 +23,55 @@ export type ChannelsRenameBodyParams = {
   displayName: string;
 }
 
+
+
```
Comment (grof) [inline src/mix/api/channels-types.ts:27]: Small eslint thing: 2 blank lines

File: src/mix/api/channels.ts [hunk 3]
```diff
@@ -30,3 +30,21 @@ export async function renameChannel(client: MixClient, requestParams: ChannelsRe
     data: body,
   })
 }
+
+export async function updateChannel(client: MixClient, requestParams: ChannelsConfigParams): Promise<MixResponse> {
+  debug('updateChannel()')
+  const {projectId, channelId, modes, color,
+    ...searchParams} = requestParams
```
Comment (grof) [inline src/mix/api/channels.ts:37]: There is no searchParams on this endpoint so it can be removed from line 37 and line 47.

File: test/commands/channels/channels-test-data.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
+/*
+   * Copyright 2022, Nuance, Inc. and its contributors.
+   * All rights reserved.
```
Comment (grof) [inline test/commands/channels/channels-test-data.ts:3]: Correct the indenting of the copyright header

File: test/commands/channels/channels-test-data.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+   *
+   * This source code is licensed under the Apache-2.0 license found in
+   * the LICENSE file in the root directory of this source tree.
+   */
+
+module.exports = {
+  rename: {
+      flags: {
+          project: '123',
```
Comment (grof) [inline test/commands/channels/channels-test-data.ts:12]: Let's stick to the indent size of 2 spaces.

File: test/commands/channels/channels-test-data.ts [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    },
+    expectedBody: {
+      channelId: 'very-long-uuid',
+      modes: ['DTMF', 'TTS', 'RICH_TEXT'],
+      color: 'LIGHT_PINK',
+    },
+    expectedResult: {
+      "channel": {
+        "id": 'very-long-uid',
+        "displayName": 'Test channel',
```
Comment (grof) [inline test/commands/channels/channels-test-data.ts:37]: Let's use quote-less keys in json.

File: test/commands/channels/configure.test.ts [hunk 7]
```diff
@@ -0,0 +1,88 @@
+/*
+  * Copyright 2022, Nuance, Inc. and its contributors.
+  * All rights reserved.
```
Comment (grof) [inline test/commands/channels/configure.test.ts:3]: Fix indenting of copyright header.

File: test/commands/channels/configure.test.ts [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+      expect(ctx.stdout).to.contain(`Channel ${td.configure.expectedResult.channel.displayName} (id: ${td.configure.expectedResult.channel.id}) updated successfully.`)
+    })
+
+  test
+    .env(testEnvData.env)
+    .stderr()
+    .command(['channels:configure',
+      '--project', td.configure.flags.project,
+      '--channel', td.configure.flags.channel,
+      '--mode', 'GARBAGE',
```
Comment (grof) [inline test/commands/channels/configure.test.ts:41]: Use a more palatable value like "INVALID_MODE"

File: src/commands/channels/configure.ts [hunk 9]
```diff
@@ -0,0 +1,139 @@
+/*
+  * Copyright 2022, Nuance, Inc. and its contributors.
+  * All rights reserved.
```
Comment (grof) [inline src/commands/channels/configure.ts:3]: Fix indentation of copyright header.

File: src/commands/channels/configure.ts [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+import {flags} from '@oclif/command'
+import makeDebug from 'debug'
+
+import * as MixFlags from '../../utils/flags'
+import * as ChannelsAPI from '../../mix/api/channels'
+import MixCommand from '../../utils/base/mix-command'
+import {ChannelModalities, ChannelsConfigParams} from '../../mix/api/channels-types'
+import {asArray} from '../../utils/as-array'
+import {MixClient, MixResponse, MixResult} from '../../mix/types'
+import chalk from 'chalk'
```
Comment (grof) [inline src/commands/channels/configure.ts:18]: Move 3rd-party import to first import section and sort.

File: src/commands/channels/configure.ts [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+import {channelColors} from '../../mix/api/utils/channel-colors'
+import {DomainOption} from '../../utils/validations'
+
+const debug = makeDebug('mix:commands:channels:configure')
+
+export default class ChannelsConfigure extends MixCommand {
+  static description = `update channel details in a project
```
Comment (grof) [inline src/commands/channels/configure.ts:25]: Insert blank line after short description (first line). Has no effect on how oclif renders it but will be consistent with the other commands.

File: src/commands/channels/configure.ts [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+    debug('transformResponse()')
+    const {data} = result as any
+    return data
+  }
+
+  outputHumanReadable(transformedData: any) {
+    debug('outputHumanReadable()')
+
+    const {displayName, id} = transformedData.channel
+    this.log(`Channel ${chalk.cyan(displayName)} (id: ${chalk.cyan(id)}) updated successfully.
```
Comment (grof) [inline src/commands/channels/configure.ts:137]: To be consistent with the other commands, the wording should be "Channel <channel> with ID <id> was updated."

File: src/commands/channels/configure.ts [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+      projectId,
+      channelId,
+      modes: modes?.map((mode: string) => mode.toUpperCase()),
+      color: color?.toUpperCase(),
+    }
+  }
+
+  captureOptions() {
+    super.captureOptions()
+    this.options.mode = this.options.mode ?
```
Comment (grof) [inline src/commands/channels/configure.ts:116]: The `mode` flag is defined with `multiple` set to `true` so I think you will get an array if that flag is provided by the user and that manipulation is not necessary.

File: src/commands/channels/configure.ts [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+      project: projectId,
+      channel: channelId,
+      mode: modes,
+      color,
+    } = options
+
+    return {
+      projectId,
+      channelId,
+      modes: modes?.map((mode: string) => mode.toUpperCase()),
```
Comment (grof) [inline src/commands/channels/configure.ts:109]: As discussed for entities, should we keep the transformations to make the data align with the endpoint closer to the API code?

File: src/commands/channels/configure.ts [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+
+  static flags = {
+    project: MixFlags.projectFlag,
+    channel: flags.string({
+      char: 'C',
+      description: 'channel id',
+      required: true,
+    }),
+    mode: {
+      ...MixFlags.modesFlag,
```
[the remark this answers is not in the document]
Comment (grof) [reply src/commands/channels/configure.ts:45]: The same modes that projects:create offers should be offered actually, in lowercase. We should also make use of oclif "options" to have free validation. However, that may affect projects:create so we need to check that.

File: src/commands/channels/configure.ts [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+    '  --channel bc40667c-e0f6-11ec-9d64-0242ac120003 \\',
+    '  --mode DTMF --mode TTS \\',
+    '  --color SALMON',
+  ]
+
+  static flags = {
+    project: MixFlags.projectFlag,
+    channel: flags.string({
+      char: 'C',
+      description: 'channel id',
```
Comment (grof) [inline src/commands/channels/configure.ts:41]: Use 'channel ID' for consistency with other commands.

File: src/commands/channels/configure.ts [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+
+    const modesOptions = options.mode?.map((mode: string) => mode.toUpperCase())
+
+    for (const mode of modesOptions ?? []) {
+      if (!(mode in modesSeen)) {
+        // TODO
+        this.error(`Unknown mode ${chalk.red(mode)}`)
+      } else if (modesSeen[mode]) {
+        // TODO
+        this.error(`Duplicate mode ${chalk.red(mode)} appears twice.`)
```
Comment (grof) [inline src/commands/channels/configure.ts:89]: Perhaps use a message that does not commit to how many times it's been seen, e.g. "Mode <mode> specified more than once". I will let you decide on the wording :) but I feel "duplicate" and "twice" sound redundant and it could be more than twice.

File: src/mix/api/channels-types.ts [hunk 18]
```diff
@@ -21,4 +23,53 @@ export type ChannelsRenameBodyParams = {
...
+export type ChannelsCreatePathParams = {
+  /** ID of the relevant project. */
+  projectId: string;
+}
+
+/** @hidden */
+export type ChannelsConfigPathParams = ChannelsCreatePathParams;
+
+// REVIEW: this should almost certainly be put in its own file
+export const ChannelModalities = [
```
Comment (grof) [inline src/mix/api/channels-types.ts:36]: If you look at how I ended up doing DataTypes in entities-types, the keys are what the UI expects and the values are what the endpoint expects. Something like that may help here.

File: src/commands/channels/configure.ts [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+    super.tryDomainOptionsValidation(options, domainOptions)
+
+    // At least one flag must be set
+    if (!options.mode && !options.color) {
+      this.error('At least one of --mode and --color must be set.')
+    }
+
+    const color = options.color?.toUpperCase()
+
+    if (color && !channelColors.slice(1).includes(color)) {
```
Comment (grof) [inline src/commands/channels/configure.ts:71]: Color and mode validations might be needed for channels:create as well. Should they move to vaidations.ts?

File: src/mix/api/channels-types.ts [hunk 20]
```diff
@@ -21,4 +23,53 @@ export type ChannelsRenameBodyParams = {
...
+  'MODE_UNSPECIFIED',
```
Comment (grof) [inline src/mix/api/channels-types.ts:37]: I just tried creating a channel through Swagger and it reject MODE_UNSPECIFIED. same for COLOR_UNSPECIFIED". So I suggest we remove those as they just complicate things. We will have to check what projects:create does with that list of colors before making that change.

**Comment by grof** [comment]
I played with the command and not providing the `color` flag results in COLOR_UNSPECIFIED being sent to the backend which results in failure. Not providing the `mode` flag results in `modes: undefined` getting set in the request body, which appears to clear modes out completely.

File: src/commands/channels/configure.ts (hunk 13 above)
Comment (oliver-daniel) [reply src/commands/channels/configure.ts:116]: I see; can you tell me more about the functionality of `asArray` then? It's a simple function of course, but I'm curious what aspect of oclif's behaviour it's trying to overcome.

File: src/commands/channels/configure.ts (hunk 14 above)
Comment (oliver-daniel) [reply src/commands/channels/configure.ts:109]: What do you mean by `keep the transformations` here?

File: src/commands/channels/configure.ts (hunk 19 above)
Comment (oliver-daniel) [reply src/commands/channels/configure.ts:71]: Good idea.

File: src/commands/channels/configure.ts (hunk 13 above)
Comment (grof) [reply src/commands/channels/configure.ts]: It was introduced in the very beginning of the project. If I remember correctly, it was first used with locales, before we got a hang of the "multiple" option on flags. The idea was to make sure an array was provided down the road as paramenter to the API function. We could probably go over the code base and remove asArray with the use of "multiple"...

File: src/commands/channels/configure.ts (hunk 14 above)
Comment (grof) [reply src/commands/channels/configure.ts]: You had made comments a few weeks back about doing the uppercasing and so on inside the API code (talking here about the code we have under the `api` directory in the project). Now, given the discussion we had yesterday about letting users use the casing they want and having `parse` fix it up, that point is likely now moot.

File: src/commands/channels/configure.ts [hunk 21]
```diff
@@ -0,0 +1,178 @@
...
+      validateChannelModeOptions(options.mode)
+    }
+
+    if (!(options.mode && options.color)) {
+      this.warn(chalk.yellow(`The V4 API currently requires both the --color and --mode flags to be set.
+This is due to a server-side error in which the default values for these flags (i.e., when not modified)
+are not recognized as valid.)`))
+    }
+  }
+
```
Comment (grof) [inline src/commands/channels/configure.ts:129]: Let's override `setRequestActionMessage` to say something like:

"Configuring channel _channelID_ in project _projectID_"

File: src/commands/channels/configure.ts [hunk 22]
```diff
@@ -0,0 +1,178 @@
...
+If either value is not supplied, the existing property on the channel will be unmodified.
+
+A note on channel modes and colors:
+For your convenience, you may enter modes and colors case-insensitively.
+In particular, you can spell any given mode/color with lowercase or capital letters,
+and with dashes ('-') in place of underscores ('_'). The value will be internally converted
+into the format the Mix API expects before the request is made. As an example, the values
+'light-pink', 'light_Pink', and 'LIGHT-PINK' are all equivalent to Mix's 'LIGHT_PINK'.
+
+Acceptable channel modes are:
```
Comment (grof) [inline src/commands/channels/configure.ts:36]: modes are already listed in the options for the `mode` flag. Are they getting repeated here for symmetry with colors?

File: src/commands/channels/configure.ts [hunk 23]
```diff
@@ -0,0 +1,178 @@
...
+audioscript
+dtmf
+interactivity
+richtext
+tts
+
+Acceptable channel colors are:
+blue
```
Comment (grof) [inline src/commands/channels/configure.ts:44]: Assuming users have a minimum width of 80 characters available to them, should we list the colors over 3 or columns and use less vertical space with the help?

File: src/commands/channels/configure.ts (hunk 22 above)
Comment (oliver-daniel) [reply src/commands/channels/configure.ts:36]: Yep -- absent a manpage or something which serves as the obvious source of truth, I figured users would want to have all the relevant information in one place, even if it's redundant. Thoughts?

File: src/commands/channels/configure.ts (hunk 23 above)
Comment (oliver-daniel) [reply src/commands/channels/configure.ts:44]: Interesting idea!

File: src/commands/channels/configure.ts [hunk 24]
```diff
@@ -0,0 +1,170 @@
...
+'light-pink', 'light_Pink', and 'LIGHT-PINK' are all equivalent to Mix's 'LIGHT_PINK'.
+
+Acceptable channel modes are:
+audioscript
+dtmf
+interactivity
+richtext
+tts
+
+Acceptable channel colors are:
```
Comment (grof) [inline src/commands/channels/configure.ts:43]: The help for channels:create has a blank line before the list of modalities but not before the list of colors. In this command here, there is no blank link in either location. I don't have a strong opinion the blank line but it should be consistent within and across channels commands.

File: src/commands/channels/configure.ts [hunk 25]
```diff
@@ -0,0 +1,170 @@
...
+
+  static examples = [
+    'mix channels:configure -P 1922  \\',
+    '  --channel bc40667c-e0f6-11ec-9d64-0242ac120003 \\',
+    '  --mode DTMF --mode TTS \\',
+    '  --color SALMON',
+  ]
+
+  static flags = {
+    project: MixFlags.projectFlag,
```
Comment (grof) [inline src/commands/channels/configure.ts:62]: Use `projectWithDefaultFlag`

File: src/commands/channels/configure.ts [hunk 26]
```diff
@@ -0,0 +1,170 @@
...
+      }
+    }
+
+    if (options.mode !== undefined) {
+      validateChannelModeOptions(options.mode)
+    }
+
+    if (!(options.mode && options.color)) {
+      this.warn(chalk.yellow(`The V4 API currently requires both the --color and --mode flags to be set.
+This is due to a server-side error in which the default values for these flags (i.e., when not modified)
```
Comment (grof) [inline src/commands/channels/configure.ts:112]: Can you rephrase to this to something like:

"Due to server-side limitations, this command currenlty requires both the --color and --mode flags to be set."

File: src/commands/channels/configure.ts [hunk 27]
```diff
@@ -0,0 +1,170 @@
...
+    debug('tryDomainOptionsValidation()')
+    super.tryDomainOptionsValidation(options, domainOptions)
+
+    // At least one flag must be set
+    if (options.mode === undefined && options.color === undefined) {
+      this.error('At least one of --mode and --color must be set.')
+    }
+
+    if (options.color !== undefined) {
+      try {
```
Comment (grof) [inline src/commands/channels/configure.ts:95]: Remove the `try/catch` as users end up getting the message from line 98.

File: test/commands/channels/configure.test.ts [hunk 28]
```diff
@@ -0,0 +1,88 @@
...
+      '--project', td.configure.flags.project,
+      '--channel', td.configure.flags.channel,
+      ...td.configure.flags.modes.map((mode: string) => `--mode=${mode}`),
+      '--color', td.configure.flags.color,
+    ])
+    .it('configures a project channel', ctx => {
+      expect(ctx.stdout).to.contain(`Channel ${td.configure.expectedResult.channel.displayName} with ID ${td.configure.expectedResult.channel.id} was updated successfully.`)
+    })
+
+  test
```
Comment (grof) [inline test/commands/channels/configure.test.ts:35]: This unit test needs to change once you remove the `try/catch` mentioned earlier.

**Comment by grof** [comment]
Follow-up review done here.

File: src/commands/channels/configure.ts [hunk 29]
```diff
@@ -0,0 +1,157 @@
...
+
+blue          brown         corn-flower
+cyan          green         grey
+indigo        light-green   light-grey  
+light-orange  light-pink    light-purple
+orange        pink          purple
+ruby          salmon        sky
+teal          yellow
+
+${chalk.bold('IMPORTANT:')} due to a current server-side limitation,
```
Comment (grof) [inline src/commands/channels/configure.ts:53]: Reuse correction made in channels:create

**Comment by grof** [comment]
Added one comment. Ready to go after that.
